packages/spatial_pooler/src/nupic/research/frameworks/spatial_pooler/spatial_pooler.py

```diff
@@ -0,0 +1,834 @@
+                                                each minicolumn looks at the overlap duty cycle of all other minicolumns within its inhibition radius 
+                                                and sets its own internal minimal acceptable duty cycle to: 
+                                                (minimum percent duty cycle before inhibition) * max(other minicolumns' duty cycles). 
+                                                if any minicolumn whose overlap duty cycle falls below this, all permanence values get boosted by 
+                                                synapse_perm_inc. raising permanences before inhibition allows a cell to search for new inputs 
+                                                when either its previously learned inputs are no longer ever active or when vast majority of inputs
+                                                are "hijacked" by other minicolumns.
+                                                default ``0.001``.
+
+        duty_cycle_period:                      period used to calculate duty cycles. higher values means it takes longer to respond in changes in 
+                                                boosting or synapses per connected cell. shorter values make it more unstable and likely to oscillate.
+                                                default ``1000``.
+
+        boost_strength:                         a number >= 0.0 that is used to control the strength of boosting. boosting increases as a function
+                                                of boost_strength. encourages minicolumns to have similar active duty cycles as their neighbors, which 
+                                                will lead to more efficient use of minicolumns. too much boosting may lead to instability of spatial
+                                                pooler outputs.
+                                                default ``0.0``.  
+
+        seed:                                   seed for numpy random generator.                    
+        '''
+
+        ''' controls number of minicolumns that are on within a local inhibition area '''
+        self.num_active_minicolumns_per_inh_area = int(num_active_minicolumns_per_inh_area)
+        self.local_density = local_density
+
+        assert (self.num_active_minicolumns_per_inh_area > 0) or (0 < self.local_density <= 0.5)
+
+
+        ''' input and and minicolumn dimensionality '''
+        self.input_dims = np.array(input_dims, ndmin=1)
+        self.minicolumn_dims = np.array(minicolumn_dims, ndmin=1)
+        self.num_inputs = np.prod(self.input_dims)
+        self.num_minicolumns = np.prod(self.minicolumn_dims)
+
+        assert (self.num_inputs > 0) and (self.num_minicolumns > 0)
+        assert (self.input_dims.size == self.minicolumn_dims.size)
+
+
+        ''' global or local inhibition '''
+        self.global_inhibition = global_inhibition
```

Comment:
We should probably check that `global_inhibition` is either `True` or `False`.

Reply:
Done.